platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot.impl;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.isolation.Isolatable;
+import org.gradle.internal.snapshot.ValueSnapshot;
+import org.gradle.internal.snapshot.ValueSnapshottingException;
+
+import javax.annotation.Nullable;
+
+import static org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSerialized;
+import static org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.javaDeserialized;
+
+@NonNullApi
+public class IsolatedArrayOfPrimitive implements Isolatable<Object> {
+    private final PrimitiveType primitiveType;
+    private final Object array;
+
+    public IsolatedArrayOfPrimitive(Object array) {
+        this(PrimitiveType.of(array.getClass()), array);
+    }
+
+    private IsolatedArrayOfPrimitive(PrimitiveType primitiveType, Object array) {
+        this.primitiveType = primitiveType;
+        this.array = array;
+    }
+
+    public static IsolatedArrayOfPrimitive fromByteArray(byte primitiveTypeCode, byte[] bytes) {
+        PrimitiveType primitiveType = PrimitiveType.fromOrdinal(primitiveTypeCode);
+        return new IsolatedArrayOfPrimitive(
+            primitiveType,
+            fromByteArray(bytes, primitiveType)
+        );
+    }
+
+    @Override
+    public void appendToHasher(Hasher hasher) {
+        throw unsupportedOperation();
+    }
+
+    @Override
+    public ValueSnapshot asSnapshot() {
+        throw unsupportedOperation();
+    }
+
+    private UnsupportedOperationException unsupportedOperation() {
+        return new UnsupportedOperationException("This operation is not supported on primitive arrays of type: " + primitiveType.arrayType);
+    }
+
+    @Override
+    public Object isolate() {
+        return primitiveType.clone(array);
+    }
+
+    @Nullable
+    @Override
+    public <S> S coerce(Class<S> type) {
+        return null;
+    }
+
+    public byte getPrimitiveTypeCode() {
+        return (byte) primitiveType.ordinal();
```

Comment:
❓ I would avoid using ordinal, since small refactoring could break deserialization. Are we sure this is safe and can't be used in some cross-version case?

Reply:
We could add a specific test forcing the current order not to change.